mixer/adapter/redisquota/redisquota.go

```diff
@@ -0,0 +1,509 @@
+
+-- update consumed token
+--------------------------------------------------------------------------------
+if reclaimed > 0 then
+  info_token = info_token - reclaimed;
+  if info_token < 0 then
+    info_token = 0
+  end
+  redis.call("HSET", key_meta, "token", info_token)
+end
+
+-- update the expiration time for automatic cleanup
+--------------------------------------------------------------------------------
+redis.call("PEXPIRE", key_meta, windowLength / 1000000)
+redis.call("PEXPIRE", key_meta, windowLength / 1000000)
+
+-- calculate available token
+--------------------------------------------------------------------------------
+local available_token = credit - info_token
+
+-- check available token and requested token
+--------------------------------------------------------------------------------
+
+if available_token <= 0 then
+  -- credit exhausted
+  return {0, 0}
+elseif available_token >= token then
+  -- increase token and bucket.token by token
+  redis.call("HINCRBY", key_meta, "token", token)
+  redis.call("HINCRBY", key_meta, "bucket.token", token)
+
+	-- save current request and set expiration time for auto cleanup
+	if (deduplicationid or '') ~= '' then
+		redis.call("HMSET", deduplicationid .. "-" .. key_meta, "token", token, "expire", timestamp + windowLength)
+		redis.call("PEXPIRE", deduplicationid .. "-" .. key_meta, windowLength / 1000000)
+	end
+
+  return {token,  windowLength}
+else
+
+  if bestEffort == 0 then
+    -- not enough token
+    return {0, 0}
+  end
+
+  -- allocate available token only
+  redis.call("HINCRBY", key_meta, "token", available_token)
+  redis.call("HINCRBY", key_meta, "bucket.token", available_token)
+
+	-- save current request and set expiration time for auto cleanup
+	if (deduplicationid or '') ~= '' then
+		redis.call("HMSET", deduplicationid .. "-" .. key_meta, "token", available_token, "expire", timestamp + windowLength)
+		redis.call("PEXPIRE", deduplicationid .. "-" .. key_meta, windowLength / 1000000)
+	end
+
+  return {available_token, windowLength}
+end
+`,
+	}
+)
+
+type (
+	builder struct {
+		quotaTypes    map[string]*quota.Type
+		adapterConfig *config.Params
+	}
+
+	handler struct {
+		// go-redis client
+		// connection pool with redis
+		client *redis.Client
+
+		// the limits we know about
+		limits map[string]*config.Params_Quota
+
+		// list of algorithm LUA scripts
+		scripts map[string]*redis.Script
+
+		// indirection to support fast deterministic tests
+		getTime func() time.Time
+
+		// logger provided by the framework
+		logger adapter.Logger
+	}
+)
+
+///////////////// Configuration Methods ///////////////
+
+func (b *builder) SetQuotaTypes(quotaTypes map[string]*quota.Type) {
+	b.quotaTypes = quotaTypes
+}
+
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.adapterConfig = cfg.(*config.Params)
+}
+
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	if len(b.adapterConfig.RedisServerUrl) == 0 {
```

Comment:
You can perform more validation here. See the list adapter's Validate function for URL validation.

Reply:
Move more validation to Validate